Merge intrinsic-tester

Stefan-Mada opened the pull request:
This adds all of the functionality for the intrinsic-tester executable, with a README in the intrinsics_tester folder. 
@regehr

zhengyang92 committed:
Support for PSRLs

zhengyang92 committed:
Support for PSRLs

zhengyang92 committed:
Support for PSRLs

zhengyang92 committed:
Support for PSRLs

zhengyang92 committed:
Support for PSRLs

zhengyang92 committed:
add a few semantics back

zhengyang92 committed:
add a few semantics back

zhengyang92 committed:
add a few semantics back

zhengyang92 committed:
add a few semantics back

zhengyang92 committed:
add a few semantics back

zhengyang92 committed:
fixed capture errors

zhengyang92 committed:
fixed capture errors

zhengyang92 committed:
fixed capture errors

zhengyang92 committed:
fixed capture errors

Stefan-Mada committed:
Pulled from upstream

Stefan-Mada committed:
Added mmx_padd_b/w/d and tests

Stefan-Mada committed:
Deleted redundant file

Stefan-Mada committed:
mmx_padd_b/w/d added with tests (#2)

add mmx_padd_b/w/d

Stefan-Mada committed:
mmx_padd_b/w/d added with tests (#2)

add mmx_padd_b/w/d

Stefan-Mada committed:
mmx_padd_b/w/d added with tests (#2)

add mmx_padd_b/w/d

Stefan-Mada committed:
mmx_padd_b/w/d added with tests (#2)

add mmx_padd_b/w/d

Stefan-Mada committed:
Merge pull request #1 from Stefan-Mada/intrinsics-init-working

Intrinsics init working

Stefan-Mada committed:
Merge upstream to get branch up to date

Stefan-Mada committed:
Added support for llvm.x86.mmx.punpckh[bw|wd|dq] and llvm.x86.mmx.punpckl[bw|wd|dq] (#3)

Stefan-Mada committed:
Added support for llvm.x86.mmx.punpckh[bw|wd|dq] and llvm.x86.mmx.punpckl[bw|wd|dq] (#3)

Stefan-Mada committed:
Added support for llvm.x86.mmx.punpckh[bw|wd|dq] and llvm.x86.mmx.punpckl[bw|wd|dq] (#3)

Stefan-Mada committed:
Added support for llvm.x86.mmx.punpckh[bw|wd|dq] and llvm.x86.mmx.punpckl[bw|wd|dq] (#3)

zhengyang92 committed:
tabs -> spaces

zhengyang92 committed:
tabs -> spaces

zhengyang92 committed:
tabs -> spaces

zhengyang92 committed:
tabs -> spaces

zhengyang92 committed:
support for psrli, psrai, pslli (#4)

zhengyang92 committed:
support for psrli, psrai, pslli (#4)

zhengyang92 committed:
support for psrli, psrai, pslli (#4)

zhengyang92 committed:
support for psrli, psrai, pslli (#4)

Stefan-Mada committed:
Merge branch 'intrinsics-init' of github.com:zhengyang92/alive2 into intrinsics-init

Stefan-Mada committed:
Intrinsics init working (#5)

* Fixed comments

Stefan-Mada committed:
Intrinsics init working (#5)

* Fixed comments

Stefan-Mada committed:
Intrinsics init working (#5)

* Fixed comments

Stefan-Mada committed:
Intrinsics init working (#5)

* Fixed comments

Stefan-Mada committed:
Merge branch 'intrinsics-init' of github.com:zhengyang92/alive2 into intrinsics-init

Stefan-Mada committed:
Compiles successfully

Stefan-Mada committed:
Cleaned up code, confirmed working on avx512 system

Stefan-Mada committed:
Not working build

Stefan-Mada committed:
Now compiles

Stefan-Mada committed:
Alive2 works

Stefan-Mada committed:
Merge pull request #2 from Stefan-Mada/intrinsics-init-driver-working

Intrinsics init driver working

Stefan-Mada committed:
Added arbitrary type integers and func prototypes

Stefan-Mada committed:
Vector code vaguely compiles, doesn't work

Stefan-Mada committed:
Vectors work!

Stefan-Mada committed:
Intrinsic call works

Stefan-Mada committed:
Added IR functions that return constant input of vector type

Stefan-Mada committed:
Added functionality for generating function call with constants

Stefan-Mada committed:
Testing functionality added

zhengyang92 committed:
fix semantics for pavgs

zhengyang92 committed:
fix semantics for pavgs

zhengyang92 committed:
fix semantics for pavgs

zhengyang92 committed:
fix semantics for pavgs

Stefan-Mada committed:
JIT compiles

Stefan-Mada committed:
JIT actually works wow

Stefan-Mada committed:
Added variable llvm function functionality

Stefan-Mada committed:
Fixed a cast

zhengyang92 committed:
wip

zhengyang92 committed:
wip

zhengyang92 committed:
wip

zhengyang92 committed:
add psigns

zhengyang92 committed:
add psigns

zhengyang92 committed:
add psigns

zhengyang92 committed:
reorder

zhengyang92 committed:
reorder

zhengyang92 committed:
reorder

zhengyang92 committed:
add all shifts

zhengyang92 committed:
add all shifts

zhengyang92 committed:
add all shifts

zhengyang92 committed:
add all pavgs

zhengyang92 committed:
add all pavgs

zhengyang92 committed:
add all pavgs

zhengyang92 committed:
fix

zhengyang92 committed:
fix

zhengyang92 committed:
fix

Stefan-Mada committed:
Added new IR function gen, added intrinsic function id DS

Stefan-Mada committed:
Added for loop functionality, crashes however

zhengyang92 committed:
refactor

zhengyang92 committed:
refactor

zhengyang92 committed:
refactor

zhengyang92 committed:
fix type bugs and memsafety bug found by valgrind

zhengyang92 committed:
fix type bugs and memsafety bug found by valgrind

zhengyang92 committed:
fix type bugs and memsafety bug found by valgrind

zhengyang92 committed:
added missing branches

zhengyang92 committed:
added missing branches

zhengyang92 committed:
added missing branches

zhengyang92 committed:
fix bad shape

zhengyang92 committed:
fix bad shape

zhengyang92 committed:
fix bad shape

zhengyang92 committed:
fix shift semantics

zhengyang92 committed:
fix shift semantics

zhengyang92 committed:
fix shift semantics

zhengyang92 committed:
add phadd

zhengyang92 committed:
add phadd

zhengyang92 committed:
add phadd

zhengyang92 committed:
phsub

zhengyang92 committed:
phsub

zhengyang92 committed:
phsub

zhengyang92 committed:
pmulhw pmulhuwø

zhengyang92 committed:
pmulhw pmulhuwø

zhengyang92 committed:
pmulhw pmulhuwø

zhengyang92 committed:
pmadd

zhengyang92 committed:
pmadd

zhengyang92 committed:
pmadd

Stefan-Mada committed:
Linker error, can't make TargetMachine

Stefan-Mada committed:
Oh my god I fixed holy ****

Stefan-Mada committed:
Cleaned up unnecessary code

Stefan-Mada committed:
Finally got the loops to work constexpr

Stefan-Mada committed:
Merge pull request #7 from zhengyang92/intrinsics-init

fix semantics for pavgs

Stefan-Mada committed:
Got compiling on avx512

Stefan-Mada committed:
Tester works to a limited degree

zhengyang92 committed:
fix two bugs found by Stefan

zhengyang92 committed:
fix two bugs found by Stefan

zhengyang92 committed:
fix two bugs found by Stefan

zhengyang92 committed:
pshufb

zhengyang92 committed:
pshufb

zhengyang92 committed:
pshufb

Stefan-Mada committed:
Fixed order of printing llvm ir

zhengyang92 committed:
fix, better style

zhengyang92 committed:
fix, better style

zhengyang92 committed:
fix, better style

Stefan-Mada committed:
Updated with dev, fixed pshufb and alive2 with 512 bit vectors

Stefan-Mada committed:
Tested all non integer intrinsics, made bug readme

Stefan-Mada committed:
Fixed README

Stefan-Mada committed:
Refactored JIT code out of main, cleaned up style

Stefan-Mada committed:
Cleaned up stlye for more consistency

Stefan-Mada committed:
Added support for vector instructions with integer operands

zhengyang92 committed:
fix 3 bugs found by Stefan

zhengyang92 committed:
fix 3 bugs found by Stefan

zhengyang92 committed:
fix 3 bugs found by Stefan

Stefan-Mada committed:
Merge branch 'dev' of github.com:zhengyang92/alive2-x86 into intrinsics-init

Stefan-Mada committed:
Updated README

Stefan-Mada committed:
Fixed README again

Stefan-Mada committed:
Made tester less verbose

zhengyang92 committed:
nit pick

zhengyang92 committed:
nit pick

zhengyang92 committed:
nit pick

Stefan-Mada committed:
Added progress bar

Stefan-Mada committed:
Added counter for outcomes that are all equal

Stefan-Mada committed:
Added printing of most common vector

Stefan-Mada committed:
Refactored randomizer code

Stefan-Mada committed:
Added multiple different options in vector random generation

Stefan-Mada committed:
Made all intrinsics have atleast 20% nontrivial testing

Stefan-Mada committed:
Fixed bug with printing 8 bit integers as chars

Stefan-Mada committed:
Merge pull request #9 from Stefan-Mada/rand-improvement

Rand improvement

Stefan-Mada committed:
Renamed alivetv-driver to intrinsics-tester

Stefan-Mada committed:
Added command line support

Stefan-Mada committed:
Updated tester README

Stefan-Mada committed:
Fixed typos in README

Stefan-Mada committed:
Fixed small typo

Stefan-Mada committed:
Added developer information

Stefan-Mada committed:
Removed unused function

Stefan-Mada committed:
Ran clang-format

Stefan-Mada committed:
Added copyright and date

Stefan-Mada committed:
Fixed progress bar not updating with range cmd argument

Stefan-Mada committed:
Merge branch 'dev' of github.com:zhengyang92/alive2-x86 into intrinsics-init

Stefan-Mada committed:
Fixed compiling without LLVM

Stefan-Mada committed:
Added jitlink to llvm libraries

Stefan-Mada committed:
Added seperate CXX flags for if tester is enabled

Stefan-Mada committed:
Updated README for building

Stefan-Mada committed:
Renamed .hpp to .h files

zhengyang92 committed:
remote deprecated mmx codes

zhengyang92 committed:
remote deprecated mmx codes

zhengyang92 commented on CMakeLists.txt:
```diff
@@ -144,6 +148,7 @@ set(UTIL_SRCS
 add_library(util STATIC ${UTIL_SRCS})
 add_dependencies(util generate_version)
 
+
```
no extra new line :)

zhengyang92 commented on intrinsics_tester/.clang-format:
```diff
@@ -0,0 +1,192 @@
+---
```
please don't include this file :)

zhengyang92 committed:
fix typo

zhengyang92 committed:
fix typo

zhengyang92 committed:
use macros for intrinsics listing

zhengyang92 committed:
use macros for intrinsics listing

zhengyang92 committed:
correction

zhengyang92 committed:
correction

zhengyang92 committed:
correction

zhengyang92 committed:
correction

zhengyang92 committed:
packuswb packsswb

zhengyang92 committed:
packuswb packsswb

zhengyang92 committed:
pack*sdw

zhengyang92 committed:
pack*sdw

zhengyang92 committed:
sv

zhengyang92 committed:
sv

zhengyang92 committed:
fix fakeshuffle

zhengyang92 committed:
fix fakeshuffle

zhengyang92 committed:
psad

zhengyang92 committed:
psad

zhengyang92 committed:
dont alow approx for synth

zhengyang92 committed:
dont alow approx for synth

zhengyang92 commented on CMakeLists.txt:
```diff
@@ -225,10 +242,13 @@ add_executable(alive-jobserver
 
 if (BUILD_LLVM_UTILS OR BUILD_TV)
-  llvm_map_components_to_libnames(llvm_libs support core irreader analysis passes transformutils)
+	llvm_map_components_to_libnames(llvm_libs ExecutionEngine orcjit support core irreader analysis passes transformutils native target jitlink)
```
no new tabs

zhengyang92 commented on intrinsics_tester/x86Intrin.h:
```diff
@@ -0,0 +1,104 @@
+
+#include "ir/instr.h"
+
+#ifndef TESTERX86INTRINBINOP_H
```
better moved before "#includes"

zhengyang92 replied:
And for code style consistency, can you change it to `#pragma once`

Stefan-Mada committed:
Fixed for code review

zhengyang92 commented:
Thanks! I'll continue read the code, will do another round soon.

zhengyang92 commented on CMakeLists.txt:
```diff
@@ -225,10 +241,13 @@ add_executable(alive-jobserver
 
 if (BUILD_LLVM_UTILS OR BUILD_TV)
-  llvm_map_components_to_libnames(llvm_libs support core irreader analysis passes transformutils)
+  llvm_map_components_to_libnames(llvm_libs ExecutionEngine orcjit support core irreader analysis passes transformutils native target jitlink)
```
Could you add `BUILD_INTRINSIC_TESTER` check to this line as well

zhengyang92 commented on CMakeLists.txt:
```diff
@@ -173,8 +177,14 @@ if (BUILD_LLVM_UTILS OR BUILD_TV)
+    intrinsics_tester/compareFunctions.cpp
+  )
+  
+  add_library(tester STATIC ${INTRINSICS_TESTER_SRCS})
```
please add `BUILD_INTRINSIC_TESTER` check to Line 185 and 187.

zhengyang92 commented on intrinsics_tester/JIT.h:
```diff
@@ -0,0 +1,162 @@
+#pragma once
+
+// Copyright (c) 2021-present The Alive2 Authors.
```
Feel free to add your name @Stefan-Mada  :)

zhengyang92 commented on intrinsics_tester/JIT.h:
```diff
@@ -0,0 +1,162 @@
+private:
+  std::unique_ptr<ExecutionSession> ES;
+
+  DataLayout DL;
```
In Alive2's convention, for header files, we uses full type names with namespaces, for example please use llvm::DataLayout rather than bare DataLayout here.

zhengyang92 commented on intrinsics_tester/commandLineUtil.h:
```diff
@@ -0,0 +1,151 @@
+  std::vector<std::string> arguments;
+
+  bool containsArg(std::string arg) {
+    if (std::find(arguments.begin(), arguments.end(), arg) != arguments.end())
```
this is equivalent to `return std::find(..)`

zhengyang92 replied:
Similar to my above comment.

zhengyang92 commented on intrinsics_tester/commandLineUtil.h:
```diff
@@ -0,0 +1,151 @@
+#pragma once
```
This commandline parser seems hacky but I am Okay with it.

zhengyang92 commented on intrinsics_tester/commandLineUtil.h:
```diff
@@ -0,0 +1,151 @@
+    for (int i = 1; i < argc; ++i) {
+      std::string arg(argv[i]);
+      // All args are made lowercase
+      std::transform(arg.begin(), arg.end(), arg.begin(), ::tolower);
```
Let's assume users knew the arguments need to be case sensitive, and we'll treat those arguments with capital letters as illegal. So remove this line

zhengyang92 commented on intrinsics_tester/commandLineUtil.h:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    setNumberOfRepetitions();
+    useDebugMode = containsArg("-d") ? true : false;
```
Condition switch is redundant here. 

`useDebugMode = containsArg("-d");`

will do the job.

zhengyang92 commented on intrinsics_tester/compareFunctions.cpp:
```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2018-present The Alive2 Authors.
```
I guess this file is copy-pasted from tools/transform.cpp. Can we reuse the compareFunction from tools/transform.cpp instead of calling this copy?

zhengyang92 commented on intrinsics_tester/intrinsics-tester.cpp:
```diff
@@ -0,0 +1,183 @@
+  // Initialize llvm module and intrinsic module
+  InitializeModule();
+
+  std::unique_ptr<llvm::orc::JIT> JITCompiler = Tester::generateIntrinsicJIT();
```
For alive2's convention, for .cpp files we prefer omit the namespace for types and put "use namespace ...;" at the top of the file.

zhengyang92 commented on intrinsics_tester/irWrapper.h:
```diff
@@ -0,0 +1,209 @@
+
+  return returnFunction->codegen();
+
+  return func;
```
Dead return. Please remove it.

zhengyang92 committed:
sync up with upstream

zhengyang92 committed:
sync up with upstream

zhengyang92 committed:
inject fksv fake call handling

zhengyang92 committed:
inject fksv fake call handling

zhengyang92 committed:
added a static function that returns return width for a intrinsic

zhengyang92 committed:
added a static function that returns return width for a intrinsic

Stefan-Mada committed:
Merge remote-tracking branch 'origin/dev' into intrinsics-init

# Conflicts:
#	CMakeLists.txt
#	ir/instr.cpp
#	llvm_util/known_fns.cpp
#	llvm_util/llvm2alive.cpp
#	tools/transform.cpp

Stefan-Mada committed:
Fixed segfault, tightly coupled tester with alive-tv definitions

Fixed a segfault caused by there being more instructions than was
hardcoded into the tester. Additionall, updated x86Intrin.cpp to
not need any updating by hand, as it will now use the ir/intrinsics.h
header to populate the array. Removed all previous "support" for MMX
instructions.

zhengyang92 committed:
fix bug in psad

zhengyang92 committed:
fix bug in psad

zhengyang92 committed:
fix packusdw / packuswb

zhengyang92 committed:
fix packusdw / packuswb

Stefan-Mada committed:
Merge branch 'dev' of github.com:minotaur-toolkit/alive2-x86 into intrinsics-init

Stefan-Mada committed:
Merge branch 'dev' of github.com:minotaur-toolkit/alive2-x86 into intrinsics-init-update

Stefan-Mada committed:
Finished code review changes

Stefan-Mada committed:
Merge branch 'intrinsics-init-update' into intrinsics-init

zhengyang92 committed:
sync with alive2

zhengyang92 committed:
Merge branch 'dev' into intrinsics-init

zhengyang92 commented:
the tester has been moved to https://github.com/minotaur-toolkit/semantic-tester , sync'ed our current alive branch and tester is good and running.

thanks @Stefan-Mada .
